support partials

skaes opened the pull request:
We have a lot of repetition in our YAML files, mostly due to cronjobs.

I attacked the problem by adding partials. Mind you, this a proposal, which is
why I haven't added documentation yet, but there are tests.  

The implemented partials system is very simplistic, although it looks
a bit like Rails templating. There's only one function supported, called
"partial", which takes two arguments: the name of the partial and an
argument hash. Partials are expected to be placed in a "partials"
subdirectory of the template directory and must have a '.yaml.erb'
suffix. Keys in the arguments hash are converted into local variables
for the partial. They are also accessible via the argument hash which is
passed in as variable 'locals'.

KnVerey commented:
Hey @skaes ! I discussed this feature with my team, and we decided it is something we'd like to accept. In addition to the use cases you've shown in your tests, I think it would be great to be able to embed a common pod spec into multiple deployments. Before I look at the details, I'd like to request that the rendering functionality be extracted into its own class now that the logic required is growing. This is also going to be necessary in https://github.com/Shopify/kubernetes-deploy/pull/228, so please coordinate with @maxlaverse (who isn't at Shopify either--it's great to have all these contributions! ❤️ ) about who will have time to do that piece first.

skaes commented:
@KnVerey Maxime and myself are working both at XING. He's in my team. We're using kubernetes-deploy in building an internal PaaS.  Which explains the many pull requests you've seen.

skaes commented:
@KnVerey also, thrilled you like the idea. Will refactor tomorrow.

skaes commented:
Can we make some progress on this pull request? I would like complete this one, before integrating  https://github.com/Shopify/kubernetes-deploy/pull/228.

KnVerey commented on lib/kubernetes-deploy/renderer.rb:
```diff
@@ -0,0 +1,62 @@
+      @id = current_sha[0...8] + "-#{SecureRandom.hex(4)}" if current_sha
+    end
+
+    def template_variables
```
Please make this method, `bind_template_variables` and `find_partial` private

skaes replied:
I don't think that improves anything.

skaes commented on README.md:
```diff
@@ -119,7 +119,44 @@ All templates must be YAML formatted. You can also use ERB. The following local
+          containers:
+          - name: cron-<%= name %>
+            image: ...
+            args: [ <%= args.map{|a| %Q("#{a}")}.join(", ") %> ]
```
You're mistaken. It generates ["cleanup"] vs. [cleanup]

skaes replied:
changed

skaes commented on README.md:
```diff
@@ -119,7 +119,44 @@ All templates must be YAML formatted. You can also use ERB. The following local
 
+#### Using partials
 
+`kubernetes-deploy` supports composing templates from so called partials in order to reduce duplication in Kubernetes YAML files. Given a template directory `DIR`, partials are searched for in `DIR/partials`and in 'DIR/../partials', in that order. They can be embedded in other ERB templates using the helper method `partial`. For example, let's assume an application needs a number of different Cro ...
```
fixed

skaes commented on README.md:
````diff
@@ -119,7 +119,44 @@ All templates must be YAML formatted. You can also use ERB. The following local
+<%= partial "cron", name: "send-mail", schedule: "0 0 * * *", args: %w(send-mails), cpu: "200m", memory: "256Mi" %>
+```
+
+Inside a partial, the parameters can be accesses as normal variables, or via a hash called `locals`. Thus, the `cron` template could like this:
````
fixed

skaes commented on lib/kubernetes-deploy/deploy_task.rb:
```diff
@@ -261,17 +256,7 @@ def indent_four(str)
-    rescue NameError => e
-      @logger.summary.add_paragraph("Error from renderer:\n  #{e.message.tr("\n", ' ')}")
-      raise FatalDeploymentError, "Template '#{filename}' cannot be rendered"
+      @renderer.render_template(filename, raw_template)
```
fine with me

skaes commented on test/fixtures/test-partials/deployment.yaml.erb:
```diff
@@ -0,0 +1,20 @@
+---
+apiVersion: extensions/v1beta1
+kind: Deployment
```
The renderer does not handle indentation. If you have an idea how to implement this feature without rewriting or monkey patching ERB, please let me know.

klautcomputing replied:
I would be really nice to come up with a solution for this. I will think about how to do this in a nice way for a bit as well.

skaes replied:
When I implemented the first version of the code, I already thought about this problem. I couldn't think of any other way to support indenting other than adding an explicit indentation parameter to the partial method, just as @KnVerey suggested, or writing a renderer. I rejected the first idea as being to error prone and clumsy and rejected the second idea as way too costly and not backwards compatible.

Beside, you can always write partials using YAML flow style, which does not require indentation:

```yaml
{
  a: 1,
  b: {
    x: 1
  }
}
```

skaes commented on test/integration/kubernetes_deploy_test.rb:
```diff
@@ -861,6 +861,24 @@ def test_invalid_context
 
+  def test_partials
+    api_version =
+      if KUBE_CLIENT_VERSION < Gem::Version.new("1.8.0")
```
Agreed and changed

skaes commented on test/integration/kubernetes_deploy_test.rb:
```diff
@@ -861,6 +861,24 @@ def test_invalid_context
+      else
+        "batch/v1beta1"
+      end
+    result = deploy_dir_without_profiling(
```
deploy_fixtures cannot be used because it tries to load the partials as YAML files, but they are not valid YAML.

skaes commented on test/integration/kubernetes_deploy_test.rb:
```diff
@@ -861,6 +861,24 @@ def test_invalid_context
+      bindings: { "cronjob_api_version" => api_version },
+      sha: nil
+    )
+    assert_deploy_success(result)
```
done

skaes commented on README.md:
````diff
@@ -119,7 +119,44 @@ All templates must be YAML formatted. You can also use ERB. The following local
+          restartPolicy: OnFailure
+```
+
+Note that partial names must have the suffix `.yaml.erb`.
````
done

skaes commented on lib/kubernetes-deploy/renderer.rb:
```diff
@@ -0,0 +1,62 @@
+      erb_binding = binding
+      bind_template_variables(erb_binding)
+      erb_binding.eval <<~EVA, __FILE__, __LINE__ + 1
+        def partial(partial, locals = {})
```
I wanted to keep the exposed interface available inside ERB as small as possible. The eval essentially reduces it to the method "partial" and "self", whereas your proposed implementation allows direct access to all instance variables, public and private methods of class Renderer.

skaes replied:
Of course, anything can be accessed via self as well, but it requires a conscious effort to do so.

skaes commented on lib/kubernetes-deploy/renderer.rb:
```diff
@@ -0,0 +1,64 @@
+    end
+
+    def find_partial(name)
+      partial_names = [name + '.yaml.erb', name + '.yml.erb']
```
I think you might have missed that the README specifies a search order for the directories it's looking in, in order to support scoping.

I would consider the existence of two partials which differ only in the suffix a user error. We could of course check for that condition (within one directory).

BTW: I don't think niceness is objectively measurable.

klautcomputing replied:
I read the readme and that's why I said that the code just implicitly reflects the that. And also wasn't talking about the suffix, but a partial in `DIR/partials` and `DIR/../partials`.

More importantly: `Dir.glob` removes 2 `each`s, a `File.join`, the need to do a `File.exists`, as well as `partial_names`, so basically the whole function. And would get rid of `@partials_dirs`. I think that's a cleaner way of doing it, which is why I suggested it.

skaes replied:
@klautcomputing how about you show me the implementation you want and I just copy and paste it? I really don't have time to waste on minor details.

klautcomputing replied:
I change the code, please find my PR here https://github.com/xing/kubernetes-deploy/pull/1

skaes commented on README.md:
```diff
@@ -120,6 +120,44 @@ All templates must be YAML formatted. You can also use ERB. The following local
 
+#### Using partials
+
+`kubernetes-deploy` supports composing templates from so called partials in order to reduce duplication in Kubernetes YAML files. Given a template directory `DIR`, partials are searched for in `DIR/partials`and in 'DIR/../partials', in that order. They can be embedded in other ERB templates using the helper method `partial`. For example, let's assume an application needs a number of different Cro ...
```
fixed

skaes commented:
@KnVerey @klautcomputing ready for further review.

KnVerey commented:
In my and @klautcomputing opinion, the ability to correctly handle indentation is an essential requirement for this feature. As I mentioned in my original comment in response to this PR, that is commonly the use case I've heard for requesting partials internally. IMO it would be considered a major bug if it did not work. You're right that this is not trivial to research and implement, and I don't know how to do it offhand either. One thing I'd definitely look into is the implementation in Rails. If you do not have the time to pursue this further, one of us will implement it branched off your work when we have the time, and then merge both pieces together. Either way, thank you for your work so far.

skaes commented:
@KnVerey @klautcomputing We have found a way to allow partials authors to use normal YAML notation when including YAML partials at arbitrary offsets from column zero. Since JSON is a subset of YAML, we simple convert the partial to a JSON string before including it in the outer template. I sincerely hope this satisfies your requirements now.

KnVerey commented on test/integration/kubernetes_deploy_test.rb:
```diff
@@ -802,6 +802,23 @@ def test_ejson_secrets_respects_no_prune_flag
   end
 
+  def test_partials
+    result = deploy_dir_without_profiling(
```
Our original discussion got lost, but I see why you can't use `deploy_fixtures` here. The best helper for this case would be `deploy_raw_fixtures("test-partials", bindings: { 'supports_partials' => 'true' })`. Incidentally, we should make a partial reference custom bindings (e.g. "supports_partials" in that command) to prove that works.

We also need another test about what happens when a referenced partial is not found (you can add at template to the 'invalid' fixture set for that purpose).

skaes commented on lib/kubernetes-deploy/renderer.rb:
```diff
@@ -0,0 +1,85 @@
+      }.merge(@bindings)
+    end
+
+    def bind_template_variables(binding, variables)
```
Note that fact that editors highlight the variable name 'binding' is a mistake on behalf of syntax highlighters. There is nothing special about the name 'binding'. But I get your point, so I changed it.

skaes commented on lib/kubernetes-deploy/renderer.rb:
```diff
@@ -0,0 +1,85 @@
+    def render_template(filename, raw_template)
+      return raw_template unless File.extname(filename) == ".erb"
+
+      binding = TemplateContext.new(self).template_binding
```
Done.

skaes commented on lib/kubernetes-deploy/renderer.rb:
```diff
@@ -0,0 +1,85 @@
+      bind_template_variables(binding, template_variables)
+
+      src = ERB.new(raw_template).result(binding)
+      if src =~ /^--- *\n/m
```
Please note that src is actually YAML code that already went though the ERB step and therefore partials inside it have been expanded. Apparently the choice of variable name was unfortunate, I'll rename it.

skaes replied:
Ouch. I added the code to the wrong method. Will fix.

skaes commented:
@KnVerey I think I have addressed all comments and known issues. I added a unit test class for testing the partials implementation, which should make it a lot simpler adding new features later on. 

A have explained the limitations of the partials system in the README and changed the partials integration test to show how a pod spec can be included via a partial.

Please let me know if I have missed to address something.

klautcomputing commented on README.md:
````diff
@@ -122,6 +122,92 @@ All templates must be YAML formatted. You can also use ERB. The following local
+  b: 2
+```
+
+This is a limitation of the current implementation.
````
Thanks for explaining this in the readme.

skaes replied:
yw

klautcomputing commented on test/fixtures/for_unit_tests/partials/factorial.yml.erb:
```diff
@@ -0,0 +1,7 @@
+value: <%= n %>
```
👍  test

skaes replied:
Yeah. I like it too. :-)

skaes commented on lib/kubernetes-deploy/renderer.rb:
```diff
@@ -0,0 +1,91 @@
+
+        template = @_renderer.find_partial(partial)
+        expanded_template = ERB.new(template, nil, '-').result(erb_binding)
+        if expanded_template =~ /^--- *\n/m
```
Done.

skaes commented:
@klautcomputing when you say rebase and merge into one commit, do you really want me to force push this onto this pull request? That would make it hard for us to merge this work with our own master and also impossible for anyone to follow the discussion on the pull request, won't it?

skaes commented:
can you please advise me how to fix the code style violations in the README?

klautcomputing commented:
Rebase and fixup is how I usually do it, especially when there are so many commits. So, just one commit that has a meaningful message about what was implemented. But if that makes things difficult for you I leave it up to you and @KnVerey

skaes commented:
Sorry, I misread the Policial messages. It complains about strings not being internationalized in the renderer unit test. Nevertheless I really don't know how to fix this. Please advise.

skaes commented on lib/kubernetes-deploy/renderer.rb:
```diff
@@ -0,0 +1,93 @@
+      bind_template_variables(erb_binding, template_variables)
+
+      ERB.new(raw_template).result(erb_binding)
+    rescue StandardError => e
```
Well, erb templates contain arbitrary ruby code, which can raise arbitrary exceptions. I don't see why NameError should be treated special here. Either rescue all non-fatal exceptions, or none at all, would be my strategy. But I'm happy to change it back to NameError.

KnVerey replied:
That's a fair point. I'm fine with `StandardError`.

skaes commented on lib/kubernetes-deploy/renderer.rb:
```diff
@@ -0,0 +1,93 @@
+        binding
+      end
+
+      def partial(partial, locals = {})
```
I agree.

skaes replied:
Variables are not inherited from an embedding partial in the current implementation. The only variables available in all templates are the ones passed in via extra bindings plus `current_sha` and `deployment_id`. I would consider it an extreme nuisance to not have those available automatically everywhere.

KnVerey replied:
I 100% agree that the extra bindings and standard variables should be available in all templates. You're right that _local_ variables are not inherited, but instance variables are. Here's how I tested that on your branch:

```yaml
# deployment.yaml.erb
<% @test = "hello" %>
<%= partial "pod", name: "pod1", args: ["echo", "foo"] %>

# pod.yaml.erb
...
    args: <%= @test %>
```

Results in:

![image](https://user-images.githubusercontent.com/4789493/35746295-5394b4f8-0814-11e8-8f69-955f69c87033.png)

skaes replied:
Yes, instance variables are available. Just like in Rails. In my my view, this is a feature.

skaes commented on lib/kubernetes-deploy/renderer.rb:
```diff
@@ -0,0 +1,93 @@
+        @_renderer.__send__(:bind_template_variables, erb_binding, variables)
+
+        template = @_renderer.__send__(:find_partial, partial)
+        expanded_template = ERB.new(template, nil, '-').result(erb_binding)
```
You are right, the error is handled on the main method. I thought that would be fine. But I'm happy to change it.

KnVerey replied:
To clarify why I'd like this, it's to be able to have the error message we show reflect the specific file the error happened in, have line number references (if any) make sense, and be able to include a dump of the relevant partial only.

skaes replied:
Changed.

skaes commented on lib/kubernetes-deploy/renderer.rb:
```diff
@@ -0,0 +1,93 @@
+        begin
+          JSON.generate(YAML.load(expanded_template))
+        rescue Psych::SyntaxError => e
+          raise "#{e.class}#{e}. Partial did not expand to valid YAML, source: #{expanded_template}"
```
Refactored the class and changed.

skaes commented on lib/kubernetes-deploy/renderer.rb:
```diff
@@ -0,0 +1,93 @@
+
+        # If we're at top level we don't need to worry about the result being
+        # included in another partial.
+        return expanded_template if expanded_template =~ /^--- *\n/m
```
The change you are proposing would invalidate templates in dozens of our projects which have been happily using the originally proposed implementation of partials system without any complaints for two months now. I do not approve of it.

My suggestion is to continue using a regexp, but protect against potential document loss by using YAML.load_stream to detect whether we missed detecting multiple documents.

KnVerey replied:
I think I've found a solution that will work for everyone:

```ruby
      docs = YAML.parse_stream(expanded_template)
      # If the partial contains multiple documents or has an explicit document header,
      # we know it cannot validly be indented in the parent, so return it immediately.
      return expanded_template unless docs.children.one? && docs.children.first.implicit
      # Make sure indentation isn't a problem by producing a single line of parseable YAML.
      # Note that JSON is a subset of YAML.
      JSON.generate(docs.children.first.to_ruby)
```

Here's evidence that `docs.children.first.implicit` is detecting exactly what we want:
```ruby
[7] pry(main)> docs = {
[7] pry(main)*   explicit_doc: YAML.parse_stream("---\nfoo: bar"),
[7] pry(main)*   implicit_doc: YAML.parse_stream("foo: bar"),
[7] pry(main)*   multiple_docs: YAML.parse_stream("foo: bar\n---\nbar: baz")
[7] pry(main)* }

[9] pry(main)> docs.each do |type, value|
[9] pry(main)*   puts "#{type} -> Multiple docs: #{!value.children.one?} / First has explicit header: #{!value.children.first.implicit}"
[9] pry(main)* end
explicit_doc -> Multiple docs: false / First has explicit header: true
implicit_doc -> Multiple docs: false / First has explicit header: false
multiple_docs -> Multiple docs: true / First has explicit header: false
```

skaes replied:
Interesting. I'll study this tomorrow.

skaes commented:
@KnVerey I think I have addressed all items of the latest code review, except sharing of instance variables during rendering. If you insist, I will change that too. I will rebase and force push once you are OK with the contents.

KnVerey commented on lib/kubernetes-deploy/renderer.rb:
```diff
@@ -0,0 +1,99 @@
+      bind_template_variables(erb_binding, template_variables)
+
+      ERB.new(raw_template, nil, '-').result(erb_binding)
+    rescue StandardError => err
```
Gave the feature another spin, including with some invalid templates, and noticed that rescuing `StandardError` both here and in `render_partial` is leading to some confusing error output when a partial is invalid (first screenshot) or includes a missing partial (second screenshot):

![image](https://user-images.githubusercontent.com/4789493/36461971-974d63e2-168e-11e8-9584-350d5874c973.png)

![image](https://user-images.githubusercontent.com/4789493/36463779-02cc0cee-169a-11e8-92ba-889027a39f13.png)


Let's not only fix that, but also provide the render sequence to help people debug. Here's one way to do that: https://github.com/Shopify/kubernetes-deploy/compare/9cf798f...partial_review3

Since logging is our main UI in this project, I'd also like to see some integration tests with thorough assertions for those error cases. That commit also shows what I have in mind there. For reference, here's the "after" output:

![image](https://user-images.githubusercontent.com/4789493/36463754-cf346a34-1699-11e8-923b-8f1b49a09762.png)

![image](https://user-images.githubusercontent.com/4789493/36463761-e2a49eea-1699-11e8-9b95-1be307f5fa45.png)

skaes commented:
Merged you proposed changes and ensured instance variables are not shared between partials. Please let me know if this is good to go or not.

skaes commented:
Rebased and squashed. Please note there are still random test failures when running against a minikube 1.8 cluster, but the they are unrelated to the PR.

klautcomputing commented on README.md:
````diff
@@ -122,6 +122,92 @@ All templates must be YAML formatted. You can also use ERB. The following local
+
+Inside a partial, parameters can be accessed as normal variables, or via a hash called `locals`. Thus, the `cron` template could like this:
+
+```erb
````
Was this forgotten/dismissed?

skaes commented:
@klautcomputing Force pushed latest change request. Might have overlooked changing it.

skaes commented:
\o/
